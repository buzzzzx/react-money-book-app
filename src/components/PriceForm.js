import React, { Component, createRef } from "react";

class PriceForm extends Component {
  static defaultProps = {
    item: {},
  };

  state = {
    validatePass: true,
    errorMessage: "",
  };

  titleInput = createRef();
  priceInput = createRef();
  dateInput = createRef();

  validateValue = (title, price, date) => {
    let canSubmit = false;
    if (title && price && date) {
      if (price < 0) {
        this.setState({
          validatePass: false,
          errorMessage: "金额必须大于零",
        });
      } else {
        this.setState({
          validatePass: true,
          errorMessage: "",
        });

        canSubmit = true;
      }
    } else {
      this.setState({
        validatePass: false,
        errorMessage: "请输入所有必选项",
      });
    }
    return canSubmit;
  };

  submitForm = (event) => {
    event.preventDefault();
    const { item, onFormSubmit } = this.props;
    const editMode = !!item.title;
    const title = this.titleInput.current.value.trim();
    const price = +this.priceInput.current.value.trim();
    const date = this.dateInput.current.value.trim();

    const canSubmit = this.validateValue(title, price, date);

    if (canSubmit) {
      if (editMode) {
        onFormSubmit({ ...item, title, price, date }, editMode);
      } else {
        onFormSubmit({ title, price, date }, editMode);
      }
    }
  };

  render() {
    const { title, price, date } = this.props.item;
    const { validatePass, errorMessage } = this.state;
    const { onCancelSubmit } = this.props;

    return (
      <form
        onSubmit={(event) => this.submitForm(event)}
        className="ml-4"
        noValidate
      >
        <div className="form-group row">
          <label className="col-md-2 col-form-label" htmlFor="title">
            标题*:
          </label>
          <div className="col-md-7">
            <input
              type="text"
              className="form-control"
              id="title"
              placeholder="请输入标题"
              defaultValue={title}
              ref={this.titleInput}
            />
          </div>
        </div>
        <div className="form-group row">
          <label className="col-md-2 col-form-label" htmlFor="price">
            金额*:
          </label>
          <div className="col-md-7">
            <input
              type="number"
              className="form-control"
              name="price"
              id="price"
              placeholder="请输入金额"
              defaultValue={price}
              ref={this.priceInput}
            />
          </div>
        </div>
        <div className="form-group row">
          <label className="col-md-2 col-form-label" htmlFor="date">
            日期*:
          </label>
          <div className="col-md-7">
            <input
              type="date"
              className="form-control"
              name="date"
              id="date"
              placeholder="请输入日期"
              defaultValue={date}
              ref={this.dateInput}
            />
          </div>
        </div>
        <button type="submit" className="btn btn-primary mr-4">
          提交
        </button>
        <button className="btn btn-secondary" onClick={onCancelSubmit}>
          取消
        </button>
        {!validatePass && (
          <div className="alert alert-danger mt-5" role="alert">
            {errorMessage}
          </div>
        )}
      </form>
    );
  }
}

export default PriceForm;
