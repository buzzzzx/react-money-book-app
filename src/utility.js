export const LIST_VIEW = "list";
export const CHART_VIEW = "chart";
