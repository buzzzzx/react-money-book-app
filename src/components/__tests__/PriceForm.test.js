import React from "react";
import { mount } from "enzyme";

import PriceForm from "../PriceForm";
import { testItem } from "../../utility";

const props = {
  onFormSubmit: jest.fn(),
  onCancelSubmit: jest.fn(),
};

const propsWithItem = {
  onFormSubmit: jest.fn(),
  onCancelSubmit: jest.fn(),
  item: testItem,
};

const getInputValue = (selector, wrapper) => {
  return wrapper.find(selector).instance().value;
};

const setInputValue = (selector, wrapper, value) => {
  wrapper.find(selector).instance().value = value;
};

let wrapperCreate;
let wrapperEdit;
let formInstance;

describe("test <PriceForm />", () => {
  beforeEach(() => {
    wrapperCreate = mount(<PriceForm {...props} />);
    wrapperEdit = mount(<PriceForm {...propsWithItem} />);
    formInstance = wrapperCreate.instance();
  });

  it("should render component to match the Snapshot", () => {
    expect(wrapperCreate).toMatchSnapshot();
    expect(wrapperEdit).toMatchSnapshot();
  });

  describe("test create mode <PriceForm />", () => {
    it("should render three inputs, one form, two buttons", () => {
      expect(wrapperCreate.find("input").length).toEqual(3);
      expect(wrapperCreate.find("form").length).toEqual(1);
      expect(wrapperCreate.find("button").length).toEqual(2);
    });

    it("three inputs should be empty", () => {
      expect(getInputValue("#title", wrapperCreate)).toEqual("");
      expect(getInputValue("#price", wrapperCreate)).toEqual("");
      expect(getInputValue("#date", wrapperCreate)).toEqual("");
    });

    it("submit form with empty input should show alert message", () => {
      wrapperCreate.find("form").simulate("submit");
      expect(formInstance.state.validatePass).toEqual(false);
      expect(wrapperCreate.find(".alert").length).toEqual(1);
      expect(props.onFormSubmit).not.toHaveBeenCalled();
    });

    it("submit form with invalid price should show alert message", () => {
      setInputValue("#title", wrapperCreate, "buy a phone");
      setInputValue("#price", wrapperCreate, "-20");
      setInputValue("#date", wrapperCreate, "2020-07-07");
      wrapperCreate.find("form").simulate("submit");
      expect(wrapperCreate.find(".alert").length).toEqual(1);
      expect(props.onFormSubmit).not.toHaveBeenCalled();
    });

    it("submit form with correct values should trigger callback with right object", () => {
      setInputValue("#title", wrapperCreate, "buy a phone");
      setInputValue("#price", wrapperCreate, "4000");
      setInputValue("#date", wrapperCreate, "2020-07-07");
      const newItem = {
        title: "buy a phone",
        price: 4000,
        date: "2020-07-07",
      };
      wrapperCreate.find("form").simulate("submit");
      expect(formInstance.state.validatePass).toEqual(true);
      expect(wrapperCreate.find(".alert").length).toEqual(0);
      expect(props.onFormSubmit).toHaveBeenCalledWith(newItem, false);
    });

    it("click cancel button should trigger right callback", () => {
      wrapperCreate.find("button").last().simulate("click");
      expect(props.onCancelSubmit).toHaveBeenCalled();
    });
  });

  describe("test edit mode with <PriceForm />", () => {
    it("render PriceForm with item should render the correct data to inputs", () => {
      expect(getInputValue("#title", wrapperEdit)).toEqual(testItem.title);
      expect(getInputValue("#price", wrapperEdit)).toEqual(
        testItem.price.toString()
      );
      expect(getInputValue("#date", wrapperEdit)).toEqual(testItem.date);
    });

    it("submit with changed value, should trigger the right callback with right object", () => {
      setInputValue("#title", wrapperEdit, "buy a HHKB");
      setInputValue("#price", wrapperEdit, 2300);
      const changedItem = {
        ...testItem,
        title: "buy a HHKB",
        price: 2300,
      };
      wrapperEdit.find("form").simulate("submit");
      expect(formInstance.state.validatePass).toEqual(true);
      expect(wrapperCreate.find(".alert").length).toEqual(0);
      expect(propsWithItem.onFormSubmit).toHaveBeenCalledWith(
        changedItem,
        true
      );
    });
  });
});
