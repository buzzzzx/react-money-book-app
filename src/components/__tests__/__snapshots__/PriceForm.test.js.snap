// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test <PriceForm /> should render component to match the Snapshot 1`] = `ReactWrapper {}`;

exports[`test <PriceForm /> should render component to match the Snapshot 2`] = `ReactWrapper {}`;
